src/null-safety/migration-guide.md

```diff
@@ -1,23 +1,399 @@
+  **Migrating an app is the same as migrating a package.**
+  Before fully migrating an app,
+  consider waiting until null safety is in a stable release.
+  In the meantime, you can learn about null safety,
+  perhaps adding migration hints to your app
+  or refactoring your app code.
+{{ site.alert.end }}
 
+[null safety]: /null-safety
+
+![Screenshot of migration tool](/null-safety/migration-tool.png)
+_This interactive tool (`dart migrate`) can simplify migration to null safety._
+
+**_[PENDING: flutter-specific stuff?
+Only things kevmoo's run into:
+--[no-]sound-null-safety (probably applies to `dart` as well)
+IDE launch options
+flutter run, flutter test flag order might be slightly different.
+pub run --enable-experiment=non-nullable test
+aka dart test --enable-experiment=non-nullable]_**
+
+## 1. Wait to migrate {#step1-wait}
+
+We strongly recommend migrating code in order, 
+with the leaves of the dependency graph being migrated first.
+For example, if package C depends on package B, which depends on package A,
+then A should be migrated to null safety first, then B, then C.
+
+![Illustration of C/B/A sentence](/null-safety/null-safety-migration-order.png){:width="454px"}<br>
+
+Although you _can_ migrate before your dependencies support null safety,
+you might have to change your code when your dependencies migrate.
+For example, if you predict that a function will take a nullable parameter but
+the package migrates it to be non-nullable,
+then passing a nullable argument becomes a compile error.
+
+
+### Switch to a beta release
+
+Start the migration process by switching to the **latest beta release**
+of either the Dart SDK or the Flutter SDK,
+and then updating your pubspec to match.
+
+How you get the latest beta release depends on whether
+you use the Flutter SDK:
+
+* If you use the Flutter SDK,
+  switch to the beta channel:
+
+  ```terminal
+$ flutter channel beta
+$ flutter upgrade
+```
+* Otherwise, download a beta release
+  from the [Dart SDK archive][].
+
+[Dart SDK archive]: /tools/sdk/archive#beta-channel
+
+Next, **update your pubspec** to the version of Dart you're using.
+For example, say you're using 2.12.0-18.0.beta:
+
+```terminal
+$ dart --version
+Dart SDK version: 2.12.0-18.0.beta (beta)... [PENDING: update version string!]
+```
+
+Update the `pubspec.yaml` file to have that version as a lower constraint:
+***[PENDING: Is it really necessary to match the release?
+Can we just use 2.12.0-0?]***
+
+```yaml
+environment:
+  sdk: '>=2.12.0-18.0.beta <3.0.0'
+```
+
+
+{{ site.alert.warn }}
+  If you return to working on production code, remember to
+  **switch back to a stable release**
+  (for example, by running `flutter channel stable`).
+{{ site.alert.end }}
+
+
+### Check dependency status
+
+Get the migration state of your package's dependencies,
+using the following command:
+
+```terminal
+$ dart pub outdated --mode=null-safety
+```
+
+If the output says **all the packages fully support null safety**,
```

Comment:
I think we have have to answer the question here of whether developers have to wait for both direct and dev (and transitive of both) dependencies.

Reply:
This is answered in a note later, but I'll move it up before the example.